### PR TITLE
setup experiments many_well

### PR DESCRIPTION
**Pull request by lollcat**
- Setup experiments for many well
- Add automatic loading of the most recent checkpoint
- Add run setup for snf on many well and gmm
- Add option to not fill buffer if checkpoint will be loaded
- Add option to specify training time limit.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Variables
+ManyWellCMD=python examples/many_well.py -m training.seed=0,1,2 # by default seeds are run in parellel
```
Comment (lollcat) [inline Makefile:2]: @VincentStimper you can use this to run all of the experiments for the many well problem. If you don't want to run all seeds in parellel in a single command then you can edit this line to instead be `training.seed=0`

File: examples/config/many_well.yaml [hunk 2]
```diff
@@ -1,52 +1,53 @@
-# with relatively small values so I can run things locally
```
Comment (lollcat) [inline examples/config/many_well.yaml:1]: This is the config that I'm using for the many well experiment - I overwrite specifics (e.g. loss type) in the Makefile. My GPUs are quite small - so yours may be better with a bigger batch size?

File: examples/setup_run.py [hunk 3]
```diff
@@ -127,34 +125,61 @@ def initial_sampler():
                                          initial_sampler=initial_sampler)
     return buffer
 
-
-def setup_trainer_and_run(cfg: DictConfig, setup_plotter: SetupPlotterFn,
+def get_load_checkpoint_dir(outer_checkpoint_dir):
```
Comment (lollcat) [inline examples/setup_run.py:128]: This ensures that the checkpoint with the most recent run, and most recent iteration is loaded.

File: examples/setup_run.py [hunk 4]
```diff
@@ -176,19 +201,28 @@ def setup_trainer_and_run(cfg: DictConfig, setup_plotter: SetupPlotterFn,
...
     else:
         buffer = None
+    if chkpt_dir is not None:
+        map_location = "cuda" if torch.cuda.is_available() and cfg.training.use_gpu else "cpu"
+        fab_model.load(os.path.join(chkpt_dir, "model.pt"), map_location)
+        opt_state = torch.load(os.path.join(chkpt_dir, 'optimizer.pt'), map_location)
+        optimizer.load_state_dict(opt_state)
+        if buffer is not None:
+            buffer.load(path=os.path.join(chkpt_dir, 'buffer.pt'))
+        print(f"\n\n****************loaded checkpoint: {chkpt_dir}*******************\n\n")
```
Comment (lollcat) [inline examples/setup_run.py:220]: This should print if you expect a checkpoint to be loaded (i.e. if I run `make mw_fab_buffer` once and then run it again, on the second time it loads the checkpoint from the first time).

File: examples/setup_run_snf.py [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            # append base samples and log probs
+            base_samples.append(x.detach().cpu())
+            base_log_w_s.append(base_log_w.detach().cpu())
+
+        base_samples = torch.cat(base_samples, dim=0)
+        base_log_w_s = torch.cat(base_log_w_s, dim=0)
+        info = {"eval_ess_flow": effective_sample_size(log_w=base_log_w_s, normalised=False).item()}
+        # note here that log prob of test set is actually rather forward KL.
+        target_info = self.target_distribution.performance_metrics(base_samples, base_log_w_s,
+                                                                 self.snf.log_prob,
```
Comment (lollcat) [inline examples/setup_run_snf.py:54]: the snf.log_prob is actually the log weights, so for the test set the "test set probability" will actually be the forward kl

File: examples/config/gmm.yaml [hunk 6]
```diff
@@ -41,15 +44,15 @@ training:
...
-  save_path:  results/many_well32/
+  save_path:  /home/laurence/work/code/FAB-TORCH/results/gmm/seed${training.seed}/snf/
 
 
 logger:
-  list_logger: true
-#  pandas_logger:
-#    save_period: 100 # how often to save the pandas dataframe as a csv
+#  list_logger: true
+  pandas_logger:
```
Comment (lollcat) [inline examples/config/gmm.yaml:55]: I use wandb for logging but I have also made this pandas logger which just writes everything to a dataframe and saves that - is that okay for you?

File: examples/config/many_well_fast.yaml [hunk 7]
```diff
@@ -0,0 +1,58 @@
+# with relatively small values so I can run things locally
```
Comment (lollcat) [inline examples/config/many_well_fast.yaml:1]: This is useful to test things quickly locally as it runs in less than a minute.

**Comment by VincentStimper** [comment]
What I also need is being able to set a runtime limit, i.e. when a checkpoint saved, the training should be stopped if by running the training until the next checkpoint the runtime limit would be exceeded. Otherwise, the jobs will just be killed after 24h on the cluster. I implemented it like this: https://github.com/lollcat/FAB-TORCH/blob/de5c84aaad8428617052ecf0b916aab9ba5b24f3/examples/aldp.py#L660
In your case, this has to be added to each trainer, right?

**Review by VincentStimper — APPROVED** [review]
Thanks for adding the changes, looks great now!
